clickhouse_connect/driver/httpclient.py

```diff
@@ -380,25 +380,47 @@ def command(self,
         return QuerySummary(self._summary(response))
 
     def _error_handler(self, response: HTTPResponse, retried: bool = False) -> None:
-        if self.show_clickhouse_errors:
+        """
+        Handles HTTP errors. Tries to be robust and provide maximum context.
+        """
+        try:
+            body = ""
+            # Always try to read the response body for context.
             try:
-                err_content = get_response_data(response)
+                # get_response_data reads body and decodes it for the error message
+                raw_body = get_response_data(response)
+                body = common.format_error(
+                    raw_body.decode(errors="backslashreplace")
+                ).strip()
             except Exception:  # pylint: disable=broad-except
-                err_content = None
-            finally:
-                response.close()
-
-            err_str = f'HTTPDriver for {self.url} returned response code {response.status}'
-            err_code = response.headers.get(ex_header)
-            if err_code:
-                err_str = f'HTTPDriver for {self.url} received ClickHouse error code {err_code}'
-            if err_content:
-                err_msg = common.format_error(err_content.decode(errors='backslashreplace'))
-                if err_msg.startswith('Code'):
-                    err_str = f'{err_str}\n {err_msg}'
-        else:
-            err_str = 'The ClickHouse server returned an error.'
+                # If we can't read or decode the body, we'll proceed without it
+                logger.warning("Failed to read error response body", exc_info=True)
+
+            # Build the error message
+            if self.show_clickhouse_errors:
+                err_code = response.headers.get(ex_header)
+                if err_code:
+                    # Prioritize the specific ClickHouse exception code if it exists.
+                    err_str = f"Received ClickHouse exception, code: {err_code}"
+                else:
+                    # Otherwise, just use the generic HTTP status
+                    err_str = f"HTTP driver received HTTP status {response.status}"
+
+                if body:
+                    # Always append the body if it exists
```

Comment:
I think we do the err_content.decode for a reason here -- in particular I remember something about UTF-8 values in the body that don't get properly shown (or error) if we don't do the decode.  There probably should have a been a test for it, but I suspect I got lazy.

Reply:
Yes, there's a specific PR for it:  https://github.com/ClickHouse/clickhouse-connect/pull/19